### ISSUE TITLE
filtering a date of a child, breaks boost_mode on the parent level

### ISSUE THREAD

**Issue by hrahal** [elastic/elasticsearch#22138]
**Elasticsearch version**: 5.0

**Plugins installed**: []

**JVM version**:

**OS version**:  OS X 10.11.6

**Description of the problem including expected versus actual behavior**:

I'm dealing with a schema that relates, companies to branches to employees,

so I'm using a parent -> child -> grand-child approach with the following mapping: 

**Steps to reproduce**:
 1. insert mapping

```
PUT 127.0.0.1:9200/elastic_issue

{
	"mappings": {
		"company": {
			"properties": {
				"name": {
					"type": "keyword"
				},
				"company_id": {
					"type": "long"
				},
				"state": {
					"type": "keyword"
				}
			}
		},
		"branch": {
			"_parent": {
				"type": "company"
			},
			"properties": {
				"opening_date": {
					"type": "date",
					"format": "epoch_millis"
				},
                                "opening_date_long": {
					"type": "long"
				}
				"name": {
					"type": "keyword"
				}
			}
		},
		"employee": {
			"_parent": {
				"type": "branch"
			},
			"properties": {
				"name": {
					"type": "keyword"
				},
				"salary": {
					"type": "long"
				},
				"bonus": {
					"type": "long"
				}
			}
		}
	}
}
```
 
2. insert data for company, branch and employee

```
127.0.0.1:9200/elastic_issue/company/1

{
	"name": "blue sky",
	"company_id": 1,
	"state": "ON"
}

127.0.0.1:9200/elastic_issue/branch/11?parent=1&routing=1

{
	"opening_date": 1481624898000,
	"opening_date_long": 1481624898000,
	"name": "zalka"
}

127.0.0.1:9200/elastic_issue/employee/xyz?parent=11&routing=1

{ 
	"name": "rick",
	"salary": 1500,
	"bonus": 300
}
```

MY requirements heavily depend on scoring on each level.

what I'm expecting to achieve is the following: 

employees that have a salary of a certain range, will take a specific weight value,
same for their bonus and the company they work for, etc..

the forumla is:

score = salary weight + bonus weight + company weight + state wight

3. query

```

{
	"query": {
		"function_score": {
			"score_mode": "sum",
			"boost_mode": "sum",
			"query": {
				"bool": {
					"must": {
						"has_child": {
							"type": "branch",
							"score_mode": "max",
							"inner_hits": {
							    "size": 1
							},
							"query": {
								"function_score": {
									"query": {
										"bool": {
											"must": {
												"has_child": {
													"type": "employee",
													"score_mode": "max",
													"inner_hits": {
														"size": 1
													},
													"query": {
														"function_score": {
															"score_mode": "sum",
															"boost_mode": "sum",
															"functions": [
																{"filter":{"range":{"salary":{"gte":0,"lt":1000}}},"weight":1},
																{"filter":{"range":{"salary":{"gte":1000,"lt":2000}}},"weight":2},
																{"filter":{"range":{"bonus":{"gte":0,"lt":33}}},"weight":2},
																{"filter":{"range":{"bonus":{"gte":33,"lt":66}}},"weight":3},
																{"filter":{"range":{"bonus":{"gte":66,"lt":100}}},"weight":1}
																
															]
														}
													}
												}
											},
											"filter": [
											    {
											        "range": {
											            "opening_date": {
											                "gte": 1481624898000
											            }
											        }
											    }
											 ]
										}
									}
								}
							}
						}
					}
				}
			},
			"functions": [
				{
					"filter": {
						"terms": {
							"company_id": [1]
						}
					},
					"weight": 2
				},
				{
					"filter": {
						"terms": {
							"state": ["ON"]
						}
					},
					"weight": 1
				}
			]
		}
	},
	"track_scores": true
}
```

at this point the employees scores are being added and passed to the parent - branch,

but when it reaches the company level the scoring breaks and the child score (branch) gets **multiplied** by the sum of the company score instead of being **summed up**

the reason behind this is the **date range filter** that is on the branch level 
```
{
	"filter": [{
		"range": {
			"opening_date": {
				"gte": 1481624898000
			}
		}
	}]
}
```

when mapping this attribute to **date** it breaks the parent score function.

however if replaced by mapping **long** mapping it works all fine,

so you can try:

```
{
	"filter": [{
		"range": {
			"opening_date_long": {
				"gte": 1481624898000
			}
		}
	}]
}
```

this works fine.

for me it seems that the date range filter is breaking the function_score "boost_mode": "sum" and forcing the child score to get **multiplied** by its parent opposed to a custom mode, "sum", "avg", etc...

I tried reproducing this on another machine and it gave the same results

using the explain API I realised: 

```
127.0.0.1:9200/elastic_issue/_search?explain=true

{
	"_shard": "[elastic_issue][3]",
	"_node": "5rSiUGoaRQi3agxKCMJ31g",
	"_index": "elastic_issue",
	"_type": "company",
	"_id": "1",
	"_score": 9,
	"_source": {
		"name": "blue sky",
		"company_id": 1,
		"state": "ON"
	},
	"_explanation": {
		"value": 9,
		"description": "function score, product of:",
		"details": [{
			"value": 3,
			"description": "A match, join value 1",
			"details": []
		}, {
			"value": 3,
			"description": "min of:",
			"details": [{
				"value": 3,
				"description": "function score, score mode [sum]",
				"details": [{
					"value": 2,
					"description": "function score, product of:",
					"details": [{
						"value": 1,
						"description": "match filter: company_id:[1 TO 1]",
						"details": []
					}, {
						"value": 2,
						"description": "product of:",
						"details": [{
							"value": 1,
							"description": "constant score 1.0 - no function provided",
							"details": []
						}, {
							"value": 2,
							"description": "weight",
							"details": []
						}]
					}]
				}, {
					"value": 1,
					"description": "function score, product of:",
					"details": [{
						"value": 1,
						"description": "match filter: state:ON",
						"details": []
					}, {
						"value": 1,
						"description": "product of:",
						"details": [{
							"value": 1,
							"description": "constant score 1.0 - no function provided",
							"details": []
						}, {
							"value": 1,
							"description": "weight",
							"details": []
						}]
					}]
				}]
			}, {
				"value": 3.4028235e+38,
				"description": "maxBoost",
				"details": []
			}]
		}]
	},
	"inner_hits": {
		"branch": {
			"hits": {
				"total": 1,
				"max_score": 3,
				"hits": [{
					"_type": "branch",
					"_id": "11",
					"_score": 3,
					"_routing": "1",
					"_parent": "1",
					"_source": {
						"opening_date": 1481624898000,
						"opening_date_long": 1481624898000,
						"name": "zalka"
					},
					"inner_hits": {
						"employee": {
							"hits": {
								"total": 1,
								"max_score": 3,
								"hits": [{
									"_type": "employee",
									"_id": "xyz",
									"_score": 3,
									"_routing": "1",
									"_parent": "11",
									"_source": {
										"name": "rick",
										"salary": 1500,
										"bonus": 300
									}
								}]
							}
						}
					}
				}]
			}
		}
	}
}
```  

in the _explanation: 

the first description is being: 

"description": "function score, product of:"

meaning its a multiplication factor although I'm forcing "boost_mode": "sum"

whereas changing the filter to "opening_date_long" *long* instead of *date*

the first description will change to: 

"description": "sum of",

works fine again.

nasty!

**Comment by jotaoncode** [comment]
+1

**Comment by clintongormley** [comment]
I've reduced this to the following:

```
PUT elastic_issue
{
  "mappings": {
    "branch": {
      "properties": {
        "opening_date": {
          "type": "date",
          "format": "epoch_millis"
        },
        "opening_date_long": {
          "type": "long"
        }
      }
    }
  }
}

PUT elastic_issue/branch/11
{
	"opening_date": 1481624898000,
	"opening_date_long": 1481624898000,
	"name": "zalka"
}

GET elastic_issue/_search?explain
{
  "query": {
    "function_score": {
      "score_mode": "sum",
      "boost_mode": "replace",
      "query": {
        "bool": {
          "must": [
            {
              "range": {
                "opening_date": {
                  "gte": 1481624898000
                }
              }
            },
            {
              "constant_score": {
                "filter": {
                  "match_all": {}
                },
                "boost": 200
              }
            }
          ]
        }
      },
      "functions": [
        {
          "filter": {
            "match_all": {}
          },
          "weight": 3
        }
      ]
    }
  }
}
```

The explanation from the above starts with:

```
        "_explanation": {
          "value": 603,
          "description": "function score, product of:",
```

even though the function score says `boost_mode: sum, score_mode: sum`.

if you change the range query to run on `opening_date` instead of `opening_date_long`, then the explanation changes to:

```
        "_explanation": {
          "value": 3,
          "description": "min of:",
          "details": [
```

which doesn't seem right either

**Comment by clintongormley** [comment]
@jimczi could you take a look at this please?

**Comment by clintongormley** [comment]
btw @hrahal - regardless of this bug, I would definitely not try to solve your problem using parent-child.  You're essentially trying to impose a relational design onto elasticsearch, and it is bound to fail. Instead, denormalise your data.

**Comment by jimczi** [comment]
This one is funky. In 5.x we added a rewrite method on the query builders (like the one for Lucene query) in order to be able to optimize some queries at the ES level. In this example you have a range query on a date field and since the index contains only one document the range query builder is rewritten in a match all docs query. When the inner query of the FunctionScoreQueryBuilder is rewritten we create a new object that contains the new query plus the other parameters of the FunctionScoreQueryBuilder. The `boost_mode` was just ignored during the copy so the default value (which is SUM) is applied.
I opened  https://github.com/elastic/elasticsearch/issues/22172 to fix this issue.
